mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java

```diff
@@ -0,0 +1,277 @@
+
+    public SystemTopicBasedSystemEventService(PulsarService pulsarService) {
+        this.pulsarService = pulsarService;
+        this.topicName = TopicName.get("pulsar/system/__mqtt_event");
+        try {
+            this.systemTopicClient = new MQTTEventSystemTopicClient(pulsarService.getClient(), topicName);
+        } catch (PulsarServerException e) {
+            throw new IllegalStateException(e);
+        }
+        this.eventCache = new ConcurrentHashMap<>();
+        this.clusterClientIds = new ConcurrentHashMap<>(2048);
+        this.listeners = new ArrayList<>();
+        this.addListener(new ConnectEventListener());
+    }
+
+    public void addListener(EventListener listener) {
+        this.listeners.add(listener);
+    }
+
+    public boolean containsConnection(Connection connection) {
+        checkReader();
+        String ip = clusterClientIds.get(connection.getClientId());
+        return ip != null && !ip.equals(NettyUtils.getIp(connection.getChannel()));
+    }
+
+    @Override
+    public CompletableFuture<Void> sendConnectEvent(Connection connection) {
+        ConnectEvent connectEvent = ConnectEvent.builder()
+                .clientId(connection.getClientId())
+                .ip(NettyUtils.getIp(connection.getChannel()))
+                .build();
+        return sendEvent(getMqttEvent(connectEvent, ActionType.INSERT))
+                .thenRun(() -> {
+                    if (log.isDebugEnabled()) {
+                        log.debug("send connect event : {}", connectEvent);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> sendDisconnectEvent(Connection connection) {
+        if (!connection.getChannel().isActive()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        ConnectEvent disconnectEvent = ConnectEvent.builder()
+                .clientId(connection.getClientId())
+                .ip(NettyUtils.getIp(connection.getChannel()))
+                .build();
+        return sendEvent(getMqttEvent(disconnectEvent, ActionType.DELETE))
+                .thenRun(() -> {
+                    if (log.isDebugEnabled()) {
+                        log.debug("send disconnect event : {}", disconnectEvent);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> sendEvent(MqttEvent event) {
+        CompletableFuture<SystemTopicClient.Writer<MqttEvent>> writerFuture = systemTopicClient.newWriterAsync();
+        return writerFuture.thenCompose(writer -> {
+            CompletableFuture<MessageId> writeFuture = ActionType.DELETE.equals(event.getActionType())
+                    ? writer.deleteAsync(event) : writer.writeAsync(event);
+            writeFuture.whenComplete((__, ex) -> {
+                if (ex != null) {
+                    log.error("[{}] send event error.", topicName, ex);
+                }
+                writer.closeAsync();
```

Comment:
Looks we need to wait for this method complete.

Reply:
As we discuss, it's ok here.